xsdvalidate/xsd_validate_activity.go

```diff
@@ -0,0 +1,226 @@
+	}
+
+	var xsdFilePath string
+
+	if u.Scheme == "file" {
+		// XSD is a local file.
+		xsdFilePath = u.Path
+	} else {
+		// XSD needs to be downloaded.
+		file, err := os.CreateTemp("", "xsdValidateActivity")
+		if err != nil {
+			return nil, err
+		}
+		defer os.Remove(file.Name())
+
+		err = downloadXsdFile(params.XSDFilePathOrURL, file.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		xsdFilePath = file.Name()
+	}
+
+	// Find matching files.
+	var filesFound []string
+
+	filesFound, err = matchingFilesRelativeToDirectory(params.DirectoryPath, params.Pattern)
+	if err != nil {
+		return nil, fmt.Errorf("XSDValidateActivity match files: %v", err)
+	}
+
+	// Check matched files.
+	var failed []string
+
+	for _, matchedFile := range filesFound {
+		// Read XML file contents.
+		p := filepath.Join(params.DirectoryPath, matchedFile)
+
+		checkFailed, err := checkXml(p, xsdFilePath)
+		if err != nil {
+			return nil, err
+		}
+
+		// Trim transfer paths from start of file paths.
+		for _, failure := range checkFailed {
+			failed = append(failed, strings.TrimPrefix(failure, params.DirectoryPath+"/"))
+		}
+	}
+
+	return &XSDValidateActivityResult{Failed: failed}, nil
+}
+
+func downloadXsdFile(url string, xsdFilePath string) error {
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return err
+	}
+
+	// Make GET request.
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	// Copy response body to file.
+	out, err := os.Create(filepath.Clean(xsdFilePath))
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func matchingFilesRelativeToDirectory(directoryPath string, pattern string) ([]string, error) {
+	if _, err := os.Stat(directoryPath); errors.Is(err, os.ErrNotExist) {
+		return nil, err
+	}
+
+	// Compile regular expression pattern.
+	r, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+
+	var subpaths []string
+
+	err = filepath.WalkDir(directoryPath, func(p string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return nil
+		}
+
+		// Skip directories.
+		if d.IsDir() {
+			return nil
+		}
+
+		// Turn path into a subpath.
+		subpath, err := filepath.Rel(directoryPath, p)
+		if err != nil {
+			return err
+		}
+
+		// If filename matches, then add to subpaths.
+		if r.MatchString(filepath.Base(subpath)) {
+			subpaths = append(subpaths, subpath)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return subpaths, nil
+}
+
+func checkXml(xmlFilePath string, xsdFilePath string) ([]string, error) {
+	var errorMessages []string
+
+	toolFilePath, err := exec.LookPath("xmllint")
+	if err != nil {
+		return nil, err
+	}
+
+	toolFilePath = filepath.Clean(toolFilePath)
+
+	xsdFilePath = filepath.Clean(xsdFilePath)
+	_, err = os.Stat(xsdFilePath)
+	if errors.Is(err, os.ErrNotExist) {
```

Comment:
I think all error types should be returned.

Reply:
(Fixed.)